iwyu_include_picker.cc

```diff
@@ -1329,13 +1329,32 @@ vector<string> IncludePicker::GetCandidateHeadersForFilepathIncludedFrom(
   // ConvertToQuotedInclude because it avoids trouble when the same
   // file is accessible via different include search-paths, or is
   // accessed via a symlink.
-  const string& quoted_include_as_typed
-      = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
-  if (!quoted_include_as_typed.empty()) {
-    vector<string>::iterator it = std::find(retval.begin(), retval.end(),
-                                            quoted_includee);
-    if (it != retval.end())
-      *it = quoted_include_as_typed;
+  vector<string>::iterator it = std::find(retval.begin(), retval.end(),
+                                          quoted_includee);
+  if (it != retval.end()) {
+      const string& quoted_include_as_typed
+          = MaybeGetIncludeNameAsWritten(including_filepath, included_filepath);
+      if (!quoted_include_as_typed.empty()) {
+          *it = quoted_include_as_typed;
+      } else {
+          string strippedIncludee = included_filepath;
+          // Directory of the file is implicitly in the search path.
+          // So if we have the full path, shorten it
+          const string implicitHeaderSearchPath =
+                  CanonicalizeHeaderSearchPath(GetParentPath(including_filepath));
+          if (StripLeft(&strippedIncludee, implicitHeaderSearchPath)) {
+              VERRS(6) << "Adding include in same directory for "
+                       << included_filepath << ": " << strippedIncludee
+                       << "\n";
+              // If the included path is absolute, we prefer the short one
+              // else we prefer the current one, which is probably relative to
+              // a search directory
+              if(IsAbsolutePath(StripQuotes(*it)))
+                  retval.insert(it, "\"" + strippedIncludee + "\"");
+              else
+                  retval.push_back("\"" + strippedIncludee + "\"");
+          }
+      }
```

Comment:
I added 3 more commits so the test suite passes with absolute paths. If you like it smaller I suggest the following:
- Merge #289 to split off the test suite changes
- Remove the 3 commits I just added from this PR and revert the above changes (-> failures when using absolute paths)
- Add another PR with the above change and the 3 commits together with a change to the test suite, to test all tests with absolute input filepaths (I got that locally by using a modified version of `TestIwyuOnRelativeFile` which simply removes the current path from the output and compares this. Works on all current tests but might need a later refactoring. One can probably just use 1 function and check for an absolute path